snapcraft/internal/lxd.py

```diff
@@ -204,18 +218,83 @@ def _ensure_container(self):
     def _setup_project(self):
         self._ensure_mount(self._project_folder, self._source)
 
+    def _get_container_address(self):
+        network = self._get_container_status()['state']['network']['eth0']
+        for address in network['addresses']:
+            if address['family'] == 'inet':
+                return address['address']
+        raise RuntimeError('No IP found for {}'.format(self._container_name))
+
     def _ensure_mount(self, destination, source):
         logger.info('Mounting {} into container'.format(source))
+        remote, container_name = self._container_name.split(':')
+        if remote != 'local':
+            self._remote_mount(destination, source)
+        else:
+            devices = self._get_container_status()['devices']
+            if destination not in devices:
+                check_call([
+                    'lxc', 'config', 'device', 'add', self._container_name,
+                    destination, 'disk', 'source={}'.format(source),
+                    'path=/{}'.format(destination)])
+
+    def _remote_mount(self, destination, source):
+        # Remove project folder in case it was used "locally" before
         devices = self._get_container_status()['devices']
-        if destination not in devices:
+        if destination in devices:
             check_call([
-                'lxc', 'config', 'device', 'add', self._container_name,
+                'lxc', 'config', 'device', 'remove', self._container_name,
                 destination, 'disk', 'source={}'.format(source),
                 'path=/{}'.format(destination)])
 
+        # Generate an SSH key and add it to the container's known keys
+        keyfile = 'id_{}'.format(self._container_name)
+        if not os.path.exists(keyfile):
+            check_call(['ssh-keygen', '-o', '-N', '', '-f', keyfile],
+                       stdout=os.devnull)
+        self._container_run(['mkdir', '-p', '/root/.ssh'])
+        self._container_run(['chmod', '700', '/root/.ssh'])
+        self._container_run(['tee', '-a',
+                             '/root/.ssh/authorized_keys'],
+                            stdin=open('{}.pub'.format(keyfile), 'r'))
+        self._container_run(['chmod', '600',
+                             '/root/.ssh/authorized_keys'])
+
+        # Use sshfs in slave mode inside SSH to reverse mount destination
+        self._install_packages(['sshfs'])
+        self._container_run(['mkdir', '-p', '/{}'.format(destination)])
+        self._container_run(['mkdir', '-p', source])
+        ssh_address = self._get_container_address()
+        logger.info('Connecting via SSH to {}'.format(ssh_address))
+        stdin1, stdout1 = os.pipe()
+        stdin2, stdout2 = os.pipe()
```

Comment:
That was my same thinking when I first reviewed it. Maybe this means that a comment in the code could be useful.